subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/worker/TestWorkerProgressListener.java

```diff
@@ -0,0 +1,128 @@
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.testing.worker;
+
+import org.gradle.api.internal.tasks.testing.DecoratingTestDescriptor;
+import org.gradle.api.internal.tasks.testing.DefaultTestClassDescriptor;
+import org.gradle.api.internal.tasks.testing.TestCompleteEvent;
+import org.gradle.api.internal.tasks.testing.TestDescriptorInternal;
+import org.gradle.api.internal.tasks.testing.TestStartEvent;
+import org.gradle.api.internal.tasks.testing.results.TestListenerInternal;
+import org.gradle.api.tasks.testing.TestDescriptor;
+import org.gradle.api.tasks.testing.TestOutputEvent;
+import org.gradle.api.tasks.testing.TestResult;
+import org.gradle.internal.Cast;
+import org.gradle.internal.logging.progress.ProgressLogger;
+import org.gradle.internal.logging.progress.ProgressLoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class TestWorkerProgressListener implements TestListenerInternal {
+
+    private final ProgressLoggerFactory factory;
+    private final ProgressLogger parentProgressLogger;
+    private final Map<String, ProgressLogger> testWorkerProgressLoggers = new HashMap<String, ProgressLogger>();
+
+    public TestWorkerProgressListener(ProgressLoggerFactory factory, ProgressLogger parentProgressLogger) {
+        this.factory = factory;
+        this.parentProgressLogger = parentProgressLogger;
+    }
+
+    @Override
+    public void started(TestDescriptorInternal testDescriptor, TestStartEvent startEvent) {
+        String description = determineTestWorkerDescription(testDescriptor);
+
+        if (description != null && !testWorkerProgressLoggers.containsKey(description)) {
+            ProgressLogger progressLogger = factory.newOperation(TestWorkerProgressListener.class, parentProgressLogger);
+            progressLogger.start(description, description);
+            testWorkerProgressLoggers.put(description, progressLogger);
+        }
+    }
+
+    @Override
+    public void completed(TestDescriptorInternal testDescriptor, TestResult testResult, TestCompleteEvent completeEvent) {
+        String description = determineTestWorkerDescription(testDescriptor);
+
+        if (description != null && testWorkerProgressLoggers.containsKey(description)) {
+            ProgressLogger progressLogger = testWorkerProgressLoggers.remove(description);
+            progressLogger.completed();
+        }
+    }
+
+    @Override
+    public void output(TestDescriptorInternal testDescriptor, TestOutputEvent event) {
+    }
+
+    /**
+     * Completes and clears registered progress loggers even if test worker crashed.
+     */
+    public void completeAll() {
+        for (ProgressLogger progressLogger : testWorkerProgressLoggers.values()) {
+            progressLogger.completed();
+        }
+
+        testWorkerProgressLoggers.clear();
+    }
+
+    private String determineTestWorkerDescription(TestDescriptorInternal testDescriptor) {
```

Comment:
Happy to move logic into a separate class. I am assuming you mean the method `determineTestWorkerDescription` and everything below? Can you explain how this will potentially be reused by build operations? I don't quite get when we'd reuse the logic.